### PR TITLE
reduce edge size

### PR DESCRIPTION
**Pull request by AviAvni**

### CODE REVIEW

**Comment by filipecosta90** [comment]
### Automated performance analysis summary

This comment was automatically generated given there is performance data available.

In summary:
- Detected a total of 10 stable tests between versions.
- Detected a total of 1 highly unstable benchmarks.
- Detected a total of 1 improvements above the improvement water line.

You can check a comparison in detail via the [grafana link](https://benchmarksrediscom.grafana.net/d//SH9_rQYGz/?var-branch=master&var-branch=perf-reduce-edge-size)

### Comparison between master and perf-reduce-edge-size.

Time Period from 30 days ago. (environment used: oss-standalone)

|                               Test Case                               |Baseline master (median obs. +- std.dev)|Comparison perf-reduce-edge-size (median obs. +- std.dev)|% change (higher-better)|            Note             |
|-----------------------------------------------------------------------|----------------------------------------|---------------------------------------------------------|------------------------|-----------------------------|
|ENTITY_COUNT                                                           | 33324 +- 26.7% UNSTABLE (7 datapoints) | 49999 +- nan%  (1 datapoints)                           |50.0%                   |UNSTABLE (very high variance)|
|EXPLICIT-EDGE-DELETION                                                 | 100 +- 0.0%  (7 datapoints)            | 100 +- nan%  (1 datapoints)                             |0.0%                    |-- no change --              |
|GRAPH500-SCALE_18-EF_16-1_HOP                                          | 33325 +- 0.0%  (7 datapoints)          | 33323 +- nan%  (1 datapoints)                           |-0.0%                   |-- no change --              |
|GRAPH500-SCALE_18-EF_16-1_HOP_MIXED_READ_65perc_WRITE_25perc_DEL_10perc| 50 +- 0.0%  (7 datapoints)             | 50 +- nan%  (1 datapoints)                              |-0.0%                   |-- no change --              |
|GRAPH500-SCALE_18-EF_16-2_HOP_MIXED_READ_75perc_WRITE_25perc           | 56 +- 2.9%  (7 datapoints)             | 59 +- nan%  (1 datapoints)                              |5.9%                    |IMPROVEMENT                  |
|IMPLICIT-EDGE-DELETION                                                 | 17 +- 0.0%  (7 datapoints)             | 17 +- nan%  (1 datapoints)                              |-0.0%                   |-- no change --              |
|INSPECT-EDGES                                                          | 2500 +- 0.0%  (7 datapoints)           | 2500 +- nan%  (1 datapoints)                            |-0.0%                   |-- no change --              |
|MIX_READ_WRITE                                                         | 10000 +- 4.4%  (7 datapoints)          | 10000 +- nan%  (1 datapoints)                           |-0.0%                   |-- no change --              |
|NODE-INDEX-LOOKUP                                                      | 33 +- 0.0%  (7 datapoints)             | 33 +- nan%  (1 datapoints)                              |-0.0%                   |-- no change --              |
|SORT_ENTITIES                                                          | 100 +- 0.0%  (7 datapoints)            | 100 +- nan%  (1 datapoints)                             |-0.0%                   |-- no change --              |
|UPDATE-BASELINE                                                        | 24999 +- 0.0%  (7 datapoints)          | 24998 +- nan%  (1 datapoints)                           |-0.0%                   |-- no change --              |
|allShortestPaths-10hop-100Kpaths                                       | 30 +- 3.4%  (7 datapoints)             | 30 +- nan%  (1 datapoints)                              |0.0%                    |-- no change --              |

File: src/graph/entities/edge.h [hunk 1]
```diff
@@ -14,30 +14,16 @@ typedef int RelationID;
...
-(Edge) {                              \
-	.entity = NULL,                   \
-	.id = INVALID_ENTITY_ID,          \
-	.relationship = NULL,             \
-	.relationID = GRAPH_NO_RELATION,  \
-	.src = NULL,                      \
-	.dest = NULL,                     \
-	.srcNodeID = INVALID_ENTITY_ID,   \
-	.destNodeID = INVALID_ENTITY_ID   \
-}
```
Comment (swilly22) [inline src/graph/entities/edge.h:28]: Why was this macro removed, don't we need a convenient way to create an initialized stack based edge ?

File: src/graph/entities/edge.c [hunk 2]
```diff
@@ -17,15 +17,29 @@ NodeID Edge_GetSrcNodeID
...
 (
 	const Edge *edge
 ) {
 	ASSERT(edge);
-	return edge->destNodeID;
+	if(edge->dest_id == INVALID_ENTITY_ID) {
+		return INVALID_ENTITY_ID;
+	}
+	if(edge->dest_id & MSB_MASK) {
+		NodeID *id = (NodeID *)(CLEAR_MSB(edge->dest_id));
```
Comment (swilly22) [inline src/graph/entities/edge.c:39]: Why do we need to deal with `CLEAR_MSB` at this point ?
I'm expecting `Edge_SetSrcNode` receive and store a "valid" node ID. MSB cleared.

File: src/constraint/constraint.c [hunk 3]
```diff
@@ -554,8 +554,8 @@ void Constraint_EnforceEdges
 
 		do {
 			Edge e;
```
Comment (swilly22) [inline src/constraint/constraint.c:556]: Shouldn't we initialize the edge using the `GE_NEW_EDGE` macro ?

File: src/execution_plan/ops/shared/create_functions.c [hunk 4]
```diff
@@ -145,26 +145,18 @@ static void _CommitEdges
...
 		AttributeSet attr = pending->edge_attributes[i];
 
-		// nodes which already existed prior to this query would
-		// have their ID set under e->srcNodeID and e->destNodeID
-		// Nodes which are created as part of this query would be
-		// saved under edge src/dest pointer
-		if(e->srcNodeID != INVALID_ENTITY_ID) srcNodeID = e->srcNodeID;
-		else srcNodeID = ENTITY_GET_ID(Edge_GetSrcNode(e));
-		if(e->destNodeID != INVALID_ENTITY_ID) destNodeID = e->destNodeID;
-		else destNodeID = ENTITY_GET_ID(Edge_GetDestNode(e));
```
Comment (swilly22) [inline src/execution_plan/ops/shared/create_functions.c:160]: Happy to see this removed!

File: src/graph/entities/edge.c [hunk 5]
```diff
@@ -17,15 +17,29 @@ NodeID Edge_GetSrcNodeID
...
 	ASSERT(edge);
-	return edge->srcNodeID;
+	if(edge->src_id == INVALID_ENTITY_ID) {
+		return INVALID_ENTITY_ID;
+	}
+	if(edge->src_id & MSB_MASK) {
+		NodeID *id = (NodeID *)(CLEAR_MSB(edge->src_id));
+		return *id;
+	}
+	return edge->src_id;
```
Comment (swilly22) [inline src/graph/entities/edge.c:27]: I don't understand this change, my hope we can simply return `edge->src_id`

File: src/graph/entities/edge.h [hunk 6]
```diff
@@ -50,17 +36,13 @@ __extension__({
 	(e)->relationID;                                                                       \
 })
 
-/* TODO: note it is possible to get into an inconsistency
- * if we set src and srcNodeID to different nodes. */
```
Comment (swilly22) [inline src/graph/entities/edge.h:54]: 👍

File: src/graph/entities/edge.h [hunk 7]
```diff
@@ -83,18 +65,6 @@ int Edge_GetRelationID
...
-Node *Edge_GetSrcNode
-(
-	Edge *e
-);
-
-// retrieve edge destination node
-Node *Edge_GetDestNode
-(
-	Edge *e
-);
```
Comment (swilly22) [inline src/graph/entities/edge.h:96]: didn't we had situations where we needed the actual src / dest nodes from a given edge ?

File: src/index/index_construct.c [hunk 8]
```diff
@@ -175,8 +175,8 @@ static void _Index_PopulateEdgeIndex
 
 		do {
 			Edge e;
```
Comment (swilly22) [inline src/index/index_construct.c:177]: Shouldn't we initialize this edge via `GE_NEW_LABELED_EDGE` ?

**Review by swilly22 — CHANGES_REQUESTED** [review]
Overall really good change!, one issue though see `edge.h/c` comments.

File: src/graph/entities/edge.c (hunk 2 above)
Comment (swilly22) [reply src/graph/entities/edge.c:39]: to avoid complexity consider switching to `ReserveNodeID` to guarantee intermediate nodes are associated with an ID.

Comment (AviAvni) [reply src/graph/entities/edge.c]: done

**Pull request merged**